src/client/QXmppOmemoManager.cpp

```diff
@@ -0,0 +1,5349 @@
+        reportFinishedResult(interface, { error });
+    } else if (QXmppOmemoIq::isOmemoIq(element)) {
+        auto omemoIq = std::make_shared<QXmppOmemoIq>();
+        omemoIq->parse(element);
+
+        auto future = d->handleReceivedStanza(*omemoIq);
+        await(future, this, [=](bool isPayloadDecrypted) mutable {
+            if (isPayloadDecrypted) {
+                // The OMEMO element is removed by creating a QXmppIq out of the
+                // QXmppOmemoIq and only its remaining elements / extensions are
+                // serialized.
+                QByteArray serializedDecryptedIq;
+                QXmlStreamWriter writer(&serializedDecryptedIq);
+                QXmppIq(*omemoIq).toXml(&writer);
+
+                QDomDocument document;
+                document.setContent(serializedDecryptedIq, true);
+
+                reportFinishedResult(interface, { document.documentElement() });
+            } else {
+                QXmpp::SendError error;
+                error.text = QStringLiteral("OMEMO message could not be decrypted");
+                error.type = QXmpp::SendError::EncryptionError;
+                reportFinishedResult(interface, { error });
+            }
+        });
+    } else {
+        reportFinishedResult(interface, { QXmppE2eeExtension::NotEncrypted() });
+    }
+
+    return interface.future();
+}
+
+QStringList QXmppOmemoManager::discoveryFeatures() const
+{
+    return {
+        QString(ns_omemo_2_devices) % "+notify"
+    };
+}
+
+bool QXmppOmemoManager::handleStanza(const QDomElement &stanza)
+{
+    if (d->isStarted && stanza.tagName() == "iq" && QXmppOmemoIq::isOmemoIq(stanza)) {
+        auto omemoIq = std::make_shared<QXmppOmemoIq>();
+        omemoIq->parse(stanza);
+
+        auto future = d->handleReceivedStanza(*omemoIq);
+        await(future, this, [=](bool isPayloadDecrypted) {
+            if (isPayloadDecrypted) {
+                // The OMEMO element is removed by creating a QXmppIq out of the
+                // QXmppOmemoIq and only its remaining elements / extensions are
+                // serialized.
+                QByteArray serializedDecryptedIq;
+                QXmlStreamWriter writer(&serializedDecryptedIq);
+                QXmppIq(*omemoIq).toXml(&writer);
+
+                QDomDocument document;
+                document.setContent(serializedDecryptedIq, true);
+
+                injectIq(document.documentElement(), omemoIq->e2eeMetadata());
+            }
+        });
+
+        return true;
+    }
+
+    return false;
+}
+
+bool QXmppOmemoManager::handleMessage(const QXmppMessage &message)
+{
+    if (d->isStarted && message.omemoElement()) {
+        auto messageCopy = std::make_shared<QXmppMessage>(message);
+
+        auto future = d->handleReceivedStanza(*messageCopy);
+        await(future, this, [=](bool isPayloadDecrypted) mutable {
+            if (isPayloadDecrypted) {
+                injectMessage(std::move(*messageCopy));
+            }
+        });
```

Comment:
What's the reason for the `shared_ptr` here? I'd suggest you call the `d->handleReceivedStanza()` with message by value (so it is copied) and return the decrypted message/iq via QFuture.